traits/02-betydb-api-access.Rmd

```diff
@@ -1,107 +1,118 @@
+
+A common way of handling private API keys is to place it in a text file in your home directory. 
+Don't put it in a project directory where it might be inadvertently shared.
+
+Here is how to find and save your API key:
+
+* click file --> new --> text file
+* copy the api key that was sent when you registered into the file
+* file --> save as '~/.betykey'
+
+For the public key, you can call this file `~/.betykey_public`. 
+
+### Ways to access API data
+
+1. Through a URL query
+2. Using the bash shell
+3. Using the R jsonlite package
+
+
+### Accessing data using a URL query
+
+## Components of a URL query
+
+* base url: `terraref.ncsa.illinois.edu/bety`
+* path to the api: `/api/beta`
+* api endpoint: `/search` or `traits` or `sites`. For BETYdb, these are the names of database tables. 
+* Query parameters: `genus=Sorghum`
+* Authentication: `key=9999999999999999999999999999999999999999` is the public key for the TERRA REF traits database. 
+
+## Constructing a URL query
+
+First, lets construct a query by putting together a URL.
+
+1. start with the database url: `terraref.ncsa.illinois.edu/bety`
+  * this url brings you to the home page
+2. Add the path to the API, `/api/beta`
+  * now we have terraref.ncsa.illinois.edu/bety/api/beta, which points to the API documentation
+3. Add the name of the table you want to query. Lets start with `variables`
+  * terraref.ncsa.illinois.edu/bety/api/beta/variables
+4. add query terms by appending a `?` and combining with `&`, for example:
+  * `key=9999999999999999999999999999999999999999`
+  * `type=trait` where the variable type is 'trait'
+  * `name=~height` where the variable name contains 'height'
+5. This is your complete query:
+  * `terraref.ncsa.illinois.edu/bety/api/beta/variables?type=trait&name=~height&key=9999999999999999999999999999999999999999`
+  * it will query all variables that are type trait and have 'height' in the name
+  * Does it return the expected values?
+  
+## Your Turn
+
+> What will the URL https://terraref.ncsa.illinois.edu/bety/api/beta/species?genus=Sorghum&key=9999999999999999999999999999999999999999 return?
+
+> Write a URL that will query the database for sites with "Field Scanner" in the name field. Hint: combine two terms with a `+` as in `Field+Scanner`
+
+What do you see? Do you think that this is all of the records? What happens if you add `&limit=none`? 
+
+
+#### Accessing data using the Shell
+
+Type the following command into a bash shell (the -O option names the output file): 
+
+```sh
+wget -O sorghum.json
+   "https://terraref.ncsa.illinois.edu/bety/api/beta/species?genus=Sorghum&key=9999999999999999999999999999999999999999"

```

Comment:
put the `\` in (`\\` doesn't work), as well as the comment.

please replace all `wget -O` with `curl -o` to make it more platform independent (e.g. work on macs where wget isn't available by default)